lib/vector-core/src/transform/mod.rs

```diff
@@ -154,6 +191,14 @@ pub trait FallibleFunctionTransform: Send + dyn_clone::DynClone + Sync {
 
 dyn_clone::clone_trait_object!(FallibleFunctionTransform);
 
+/// Similar to `FunctionTransform`, but with a map of outputs to handle routing events without
+/// having to clone them.
+pub trait DispatchFunctionTransform: Send + dyn_clone::DynClone + Sync {
+    fn transform(&mut self, output: &mut Vec<(Option<String>, Event)>, event: Event);
```

Comment:
As you probably noticed, this is quite similar to the `SynchronousTransform` that I implemented in #10265 😄 

Obviously I'm biased, but I would prefer we go with something closer to my version for a couple of reasons:
1. Here we're directly exposing the structure of a relatively complex output buffer, which makes it both more difficult to use and harder to change down the line
2. Having an owned `String` for each event that we output is likely going to be the source of a lot of copies, which won't be good for performance or memory pressure.
3. We don't have a good way to enforce that transforms are only writing to known outputs. Neither solution is as good as a compile time check, but my `TransformOutputsBuf` at least tries to panic as soon as possible on a buggy implementation.